drivers: uart_ns16550: device config_info content mutated

The device config_info structure is intended to be read-only, and so
the generic pointer stored in struct device is a pointer-to-const.
The uart_ns16550 driver removes this qualifier when casting the pointer to
the device-specific type.  This removal is a violation of MISRA 11.8
as it evokes undefined behavior.

When the qualifier is preserved the driver does not build, because it
modifies the object content at runtime.  Consequently the fix for this
driver has been removed from #25248.

https://github.com/pabigot/zephyr/commits/nordic/issue/24745e provides
the changes necessary to preserve const correctness.  When applied it
produces the following diagnostic:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/serial/uart_ns16550.c: In function 'uart_ns16550_configure':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/serial/uart_ns16550.c:333:25: error: assignment of member 'port' in read-only object
  333 |   dev_cfg->devconf.port = pcie_get_mbar(dev_cfg->pcie_bdf, 0);
      |                         ^
```

The driver must be updated to resolve this and incorporate the change
from the referenced branch.

--- a/drivers/serial/uart_ns16550.c
+++ b/drivers/serial/uart_ns16550.c
@@ -202,7 +202,7 @@ BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "NS16550(s) in DT need CONFIG_PCIE");
 /* convenience defines */
 
 #define DEV_CFG(dev) \
-	((struct uart_ns16550_device_config * const) \
+	((const struct uart_ns16550_device_config * const) \
 	 (dev)->config_info)
 #define DEV_DATA(dev) \
 	((struct uart_ns16550_dev_data_t *)(dev)->driver_data)
@@ -333,7 +333,7 @@ static int uart_ns16550_configure(struct device *dev,
 				const struct uart_config *cfg)
 {
 	struct uart_ns16550_dev_data_t * const dev_data = DEV_DATA(dev);
-	struct uart_ns16550_device_config * const dev_cfg = DEV_CFG(dev);
+	const struct uart_ns16550_device_config * const dev_cfg = DEV_CFG(dev);
 
 	u8_t mdc = 0U;
 

